dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/persistence/datastore/FileDataStoreHandler.kt

```diff
@@ -0,0 +1,289 @@
+        val dataBlock = getDataBlock(data, serializer)
+
+        if (lastUpdateBlock == null || versionCodeBlock == null || dataBlock == null) return
+
+        writeToFile(
+            dataStoreFile,
+            lastUpdateBlock + versionCodeBlock + dataBlock
+        )
+    }
+
+    @WorkerThread
+    override fun <T : Any> read(
+        dataStoreFileName: String,
+        featureName: String,
+        deserializer: Deserializer<String, T>,
+        version: Int
+    ): T? {
+        val dataStoreDirectory = dataStoreFileHelper.getDataStoreDirectory(
+            sdkInstanceId = sdkInstanceId,
+            featureName = featureName,
+            folderName = DATASTORE_FOLDER_NAME.format(Locale.US, version),
+            storageDir = storageDir
+        )
+
+        val datastoreFile = dataStoreFileHelper.getDataStoreFile(
+            dataStoreDirectory = dataStoreDirectory,
+            dataStoreFileName = dataStoreFileName
+        )
+
+        if (!datastoreFile.existsSafe(internalLogger)) {
+            return null // no datastore file found
+        }
+
+        return readFromDataStoreFile(datastoreFile, deserializer, fileTLVBlockReader, version)
+    }
+
+    @WorkerThread
+    private fun writeToFile(dataStoreFile: File, data: ByteArray) {
+        fileReaderWriter.writeData(
+            file = dataStoreFile,
+            data = data,
+            append = false
+        )
+    }
+
+    @WorkerThread
+    private fun <T : Any> getDataBlock(
+        data: T,
+        serializer: Serializer<T>
+    ): ByteArray? {
+        val serializedData = serializer.serialize(data)?.toByteArray()
+
+        if (serializedData == null) {
+            logFailedToSerializeDataError()
+            return null
+        }
+
+        val dataBlock = TLVBlock(
+            type = TLVBlockType.DATA,
+            data = serializedData
+        )
+
+        return dataBlock.serialize()
+    }
+
+    @WorkerThread
+    private fun getLastUpdateDateBlock(): ByteArray? {
+        val now = System.currentTimeMillis()
+        val lastUpdateDateByteArray = now.toByteArray()
+        val lastUpdateDateBlock = TLVBlock(
+            type = TLVBlockType.LAST_UPDATE_DATE,
+            data = lastUpdateDateByteArray
+        )
+
+        return lastUpdateDateBlock.serialize()
+    }
+
+    @WorkerThread
+    private fun getVersionCodeBlock(): ByteArray? {
+        val versionCodeByteArray = CURRENT_DATASTORE_VERSION.toByteArray()
+        val versionBlock = TLVBlock(
+            type = TLVBlockType.VERSION_CODE,
+            data = versionCodeByteArray
+        )
+
+        return versionBlock.serialize()
+    }
+
+    @WorkerThread
+    @Suppress("ReturnCount")
+    private fun <T : Any> readFromDataStoreFile(
+        datastoreFile: File,
+        deserializer: Deserializer<String, T>,
+        fileTLVBlockReader: FileTLVBlockReader,
+        requestedVersion: Int
+    ): T? {
+        val tlvBlocks = fileTLVBlockReader.all(datastoreFile)
```

Comment:
maybe `read` is a better name here? Kotlin is using `all` method to check if all items satisfy criteria.

```suggestion
        val tlvBlocks = fileTLVBlockReader.read(datastoreFile)
```